zerver/actions/message_flags.py

```diff
@@ -262,6 +263,22 @@ def do_update_message_flags(
     flagattr = getattr(UserMessage.flags, flag)
 
     with transaction.atomic(savepoint=False):
+        if flag == "read" and operation == "remove":
+            subscribed_recipient_ids = get_subscribed_stream_recipient_ids_for_user(user_profile)
+
+            messages_in_unsubscribed_streams = (
+                Message.objects.select_related("recipient")
+                .filter(id__in=messages, recipient__type=Recipient.STREAM)
+                .exclude(recipient_id__in=subscribed_recipient_ids)
+                .values_list("id", flat=True)
+            )
+
+            messages = [
+                message_id
+                for message_id in messages
+                if message_id not in set(messages_in_unsubscribed_streams)
```

Comment:
This algorithm looks correct to me. @andersk FYI.

The other thing that I think we wanted to do is return some details about the set of streams to the HTTP request so that we can perhaps display those details to the end user. But I think that'll require some nontrivial plumbing and also thinking about whether we want just the set of streams, or also the number.